lib/TrackerManager.py

```diff
@@ -0,0 +1,230 @@
+
+cached_tracker = None
+cached_hashed_values = {}
+refresh_hashed_values = False
+# swdc_tracker will initialize on the first use of it (editor activated event)
+# and use a cached instance for every subsequent call
+def swdc_tracker(use_cache = True):
+	global cached_tracker
+
+	if(cached_tracker and use_cache):
+		return cached_tracker
+	else:
+		response = requestIt('GET', '/plugins/config', None, None)
+		config = json.loads(response.read().decode('utf-8'))
+		e = Emitter(config['tracker_api'])
+		tracker = Tracker(e, namespace='CodeTime', app_id='swdc-sublime')
+		cached_tracker = tracker
+		return tracker
+
+def tracker_enabled():
+	return getValue("software_telemetry_on", True)
+
+def track_codetime_event(**kwargs):
+	if tracker_enabled():
+		event_json = codetime_payload(**kwargs)
+		context = build_context(**kwargs)
+		swdc_tracker().track_self_describing_event(event_json, context)
+		if(refresh_hashed_values):
+			fetch_user_hashed_values()
+
+def track_editor_action(**kwargs):
+	if tracker_enabled():
+		event_json = editor_action_payload(**kwargs)
+		context = build_context(**kwargs)
+		response = swdc_tracker().track_self_describing_event(event_json, context)
+		if(refresh_hashed_values):
+			fetch_user_hashed_values()
+
+def track_ui_interaction(**kwargs):
+	if tracker_enabled():
+		event_json = ui_interaction_payload(**kwargs)
+		context = build_context(**kwargs)
+		swdc_tracker().track_self_describing_event(event_json, context)
+		if(refresh_hashed_values):
+			fetch_user_hashed_values()
+
+def build_context(**kwargs):
+	ctx = []
+
+	if 'jwt' in kwargs:
+		ctx.append(auth_payload(**kwargs))
+
+	if 'start_time' in kwargs and 'end_time' in kwargs:
+		ctx.append(codetime_payload(**kwargs))
+
+	if 'entity' in kwargs:
+		ctx.append(editor_action_payload(**kwargs))
+
+	if 'interaction_type' in kwargs:
+		ctx.append(ui_interaction_payload(**kwargs))
+
+	if 'element_name' in kwargs and 'element_location' in kwargs:
+		ctx.append(ui_element_payload(**kwargs))
+
+	if 'project_name' in kwargs and 'project_directory' in kwargs:
+		ctx.append(project_payload(**kwargs))
+
+	if 'repo_identifier' in kwargs:
+		ctx.append(repo_payload(**kwargs))
+
+	if 'file_name' in kwargs and 'file_path' in kwargs:
+		ctx.append(file_payload(**kwargs))
+
+	if 'plugin_id' in kwargs and 'plugin_version' in kwargs:
+		ctx.append(plugin_payload(**kwargs))
+
+	return ctx
+
+def codetime_payload(**kwargs):
+	return SelfDescribingJson(
+		'iglu:com.software/codetime/jsonschema/1-0-1',
+		{
+			'keystrokes': kwargs['keystrokes'],
+			'chars_added': kwargs['chars_added'],
+			'chars_deleted': kwargs['chars_deleted'],
+			'chars_pasted': kwargs['chars_pasted'],
+			'pastes': kwargs['pastes'],
+			'lines_added': kwargs['lines_added'],
+			'lines_deleted': kwargs['lines_deleted'],
+			'start_time': datetime.utcfromtimestamp(int(kwargs['start_time'])).isoformat(),
+			'end_time': datetime.utcfromtimestamp(int(kwargs['start_time'])).isoformat()
```

Comment:
Should this be: `'end_time': datetime.utcfromtimestamp(int(kwargs['end_time'])).isoformat()`?

Reply:
ooh good catch